File: app/src/main/java/com/example/rider/RegisterActivity.java
package com.example.rider;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import android.Manifest;
import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.nfc.Tag;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.provider.MediaStore;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.HashMap;
import java.util.Map;

public class RegisterActivity extends AppCompatActivity {
    public static final String TAG = "TAG";
    Button register, cancel;
    TextView selectProfile;
    ImageView profileImage;

    EditText inputFirstName, inputLastName, inputEmail, inputAddress, inputPhoneNumber, inputPassword;
    String emailPattern = "^(?=.{1,64}@)[A-Za-z0-9_-]+(\\.[A-Za-z0-9_-]+)@"
            + "[^-][A-Za-z0-9-]+(\\.[A-Za-z0-9-]+)(\\.[A-Za-z]{2,})$";
    ProgressDialog progressDialog;

    RadioGroup radioGroup;
    RadioButton rb_volunteer, rb_student;

    FirebaseAuth fAuth;
    FirebaseUser fUser;
    FirebaseFirestore fStore;
    String userID;

    Uri setImageURI;
    private static final int IMAGE_PICK_CODE =1000;
    private static final int PERRMISSION_CODE =1001;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);

        inputFirstName = findViewById(R.id.register_firstname_id);
        inputLastName = findViewById(R.id.register_lastname_id);
        inputEmail = findViewById(R.id.register_email_id);
        inputPassword = findViewById(R.id.register_password_id);
        inputAddress = findViewById(R.id.register_address_id);
        inputPhoneNumber =  findViewById(R.id.register_phone_id);


        register = findViewById(R.id.register_registerBtn_id);
        cancel = findViewById(R.id.register_cancelBtn_id);
        selectProfile =findViewById(R.id.register_selectProfile_id);
        profileImage =  findViewById(R.id.registerimageSelect_id);

        radioGroup = findViewById(R.id.radioGroup_Id);
        rb_student = findViewById(R.id.rb_student_id);
        rb_volunteer = findViewById(R.id.rb_volunteer_id);

        fAuth =  FirebaseAuth.getInstance();
        fUser = fAuth.getCurrentUser();
        fStore =FirebaseFirestore.getInstance();

        cancel.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                finish();
            }
        });



        register.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                performAuth();
                checkButton();
            }
        });

   selectProfile.setOnClickListener(new View.OnClickListener() {
       @Override
       public void onClick(View view) {
           selectImage();
       }
   });


}

    private void checkButton() {
        switch (radioGroup.getCheckedRadioButtonId()) {
            case R.id.rb_student_id:
                Intent Aintent = new Intent(getApplicationContext(), StudentLogin.class);
                startActivity(Aintent);
                break;

            case R.id.rb_volunteer_id:
                Intent Cintent = new Intent(getApplicationContext(), VolunteerLogin.class);
                startActivity(Cintent);
                break;
        }
    }

    private void performAuth() {
        String email = inputEmail.getText().toString().trim();
        String password = inputPassword.getText().toString().trim();
        String firstName = inputFirstName.getText().toString().trim();
        String lastName = inputLastName.getText().toString().trim();
        String address = inputAddress.getText().toString().trim();
        String phoneNumber = inputPhoneNumber.getText().toString().trim();

        if (TextUtils.isEmpty(firstName)) {
            inputFirstName.setError("FirstName is required");
            return;
        }

        if (TextUtils.isEmpty(lastName)) {
            inputLastName.setError("LastName is required");
            return;
        }

        if (TextUtils.isEmpty(email)) {
            inputEmail.setError("Email is required");
            return;
        }
        if (email.matches(emailPattern)){
            inputEmail.setError("Email is badly formatted");
            return;
        }
        if (TextUtils.isEmpty(password)) {
            inputPassword.setError("Password is required");
            return;
        }

        if (TextUtils.isEmpty(address)) {
            inputAddress.setError("Address is required");
            return;
        }
        if (password.length() < 6) {
            inputPassword.setError("Password must be >= 8 characters");
            return;
        }
        if (phoneNumber.length() < 10) {
            inputPhoneNumber.setError("Password must be = 10 digits");
            return;
        }


//        else {
//            progressDialog.setMessage("Plz wait while registration");
//            progressDialog.setTitle("Registration");
//            progressDialog.setCanceledOnTouchOutside(false);
//            progressDialog.show();
//        }

        fAuth.createUserWithEmailAndPassword(email, password).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
            @Override
            public void onComplete(@NonNull Task<AuthResult> task) {
                if (task.isSuccessful()){
                    Toast.makeText(RegisterActivity.this, "User Created", Toast.LENGTH_SHORT).show();
                    userID = fAuth.getCurrentUser().getUid();
                    DocumentReference documentReference = fStore.collection("users").document(userID);
//                    Intent intent = new Intent(RegisterActivity.this, LoginActivity.class);
                    Map<String, Object> user = new HashMap<>();
                    user.put("firstName", firstName);
                    user.put("lastName", lastName);
                    user.put("email", email);
                    user.put("phoneNumber", phoneNumber);
                    user.put("address", address);
                    documentReference.set(user).addOnSuccessListener(new OnSuccessListener<Void>() {
                        @Override
                        public void onSuccess(Void aVoid) {
                            Log.d(TAG, "onSuccess: user Profile is created" + userID);
                        }
                    });
//                    startActivity(intent);
                }
                else {
                    Toast.makeText(RegisterActivity.this, "Error !" + task.getException().getMessage(), Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    private void selectImage() {
        final  CharSequence [] options = {"Take Photo", "Choose from Gallery","cancel"};
        AlertDialog.Builder builder = new AlertDialog.Builder(RegisterActivity.this);
        builder.setTitle("Add Photo!");
        builder.setItems(options, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                if (options[i].equals("Take Photo"))
                {
                    Intent intent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
                    File f = new File(android.os.Environment.getExternalStorageDirectory(), "temp.jpg");
                    intent.putExtra(MediaStore.EXTRA_OUTPUT, Uri.fromFile(f));
                    startActivityForResult(intent, 1);
                }
                else if (options[i].equals("Choose from Gallery"))
                {
                    Intent intent = new   Intent(Intent.ACTION_PICK,android.provider.MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
                    startActivityForResult(intent, 2);
                }
                else if (options[i].equals("Cancel")) {
                    dialogInterface.dismiss();
                }
            }
        });
        builder.show();
    }
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK) {
            if (requestCode == 1) {
                File f = new File(Environment.getExternalStorageDirectory().toString());
                for (File temp : f.listFiles()) {
                    if (temp.getName().equals("temp.jpg")) {
                        f = temp;
                        break;
                    }
                }
                try {
                    Bitmap bitmap;
                    BitmapFactory.Options bitmapOptions = new BitmapFactory.Options();
                    bitmap = BitmapFactory.decodeFile(f.getAbsolutePath(),
                            bitmapOptions);
                    profileImage.setImageBitmap(bitmap);
                    String path = android.os.Environment
                            .getExternalStorageDirectory()
                            + File.separator
                            + "Phoenix" + File.separator + "default";
                    f.delete();
                    OutputStream outFile = null;
                    File file = new File(path, String.valueOf(System.currentTimeMillis()) + ".jpg");
                    try {
                        outFile = new FileOutputStream(file);
                        bitmap.compress(Bitmap.CompressFormat.JPEG, 85, outFile);
                        outFile.flush();
                        outFile.close();
                    } catch (FileNotFoundException e) {
                        e.printStackTrace();
                    } catch (IOException e) {
                        e.printStackTrace();
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            } else if (requestCode == 2) {
                Uri selectedImage = data.getData();
                String[] filePath = { MediaStore.Images.Media.DATA };
                Cursor c = getContentResolver().query(selectedImage,filePath, null, null, null);
                c.moveToFirst();
                int columnIndex = c.getColumnIndex(filePath[0]);
                String picturePath = c.getString(columnIndex);
                c.close();
                Bitmap thumbnail = (BitmapFactory.decodeFile(picturePath));
                profileImage.setImageBitmap(thumbnail);
            }
        }
    }

}

File: app/src/main/java/com/yatra/yatraapp/ui/RegisterActivity.kt
package com.yatra.yatraapp.ui

import android.app.ProgressDialog
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.provider.MediaStore
import android.text.TextUtils
import android.util.Log
import android.util.Patterns
import android.widget.ArrayAdapter
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.FileProvider
import androidx.core.widget.doOnTextChanged
import coil.load
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.firebase.auth.ktx.auth
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import com.google.firebase.storage.ktx.storage
import com.google.firebase.storage.ktx.storageMetadata
import com.mapbox.search.*
import com.mapbox.search.result.SearchResult
import com.mapbox.search.result.SearchSuggestion
import com.yatra.yatraapp.R
import com.yatra.yatraapp.databinding.ActivityRegisterBinding
import com.yatra.yatraapp.model.User
import com.yatra.yatraapp.utils.showShortToast
import java.io.File
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*

class RegisterActivity : AppCompatActivity() {
    private var binding: ActivityRegisterBinding? = null

    private var userType: Int = User.TYPE_STUDENT

    private lateinit var currentProfilePhotoUri: Uri
    private lateinit var currentLegalPhotoUri: Uri
    private var isProfile: Boolean? = null

    private lateinit var location: String
    private var latitude: Double = 0.0
    private var longitude: Double = 0.0

    private lateinit var searchEngine: SearchEngine
    private lateinit var searchRequestTask: SearchRequestTask

    private val searchCallback = object : SearchSelectionCallback {

        override fun onSuggestions(
            suggestions: List<SearchSuggestion>,
            responseInfo: ResponseInfo
        ) {
            if (suggestions.isEmpty()) {
                "No suggestions found".showShortToast(this@RegisterActivity)
            } else {
                val adapter = ArrayAdapter(
                    this@RegisterActivity,
                    android.R.layout.select_dialog_item,
                    suggestions.map { it.name }
                )

                binding?.registerAddressId?.setAdapter(adapter)
                binding?.registerAddressId?.setOnItemClickListener { _, _, pos, _ ->
                    searchRequestTask = searchEngine.select(suggestions[pos], this)
                }
            }
        }

        override fun onResult(
            suggestion: SearchSuggestion,
            result: SearchResult,
            responseInfo: ResponseInfo
        ) {
            location = result.name
            latitude = result.coordinate?.latitude()!!
            longitude = result.coordinate?.longitude()!!
            "$location $latitude $longitude".showShortToast(this@RegisterActivity)
        }

        override fun onCategoryResult(
            suggestion: SearchSuggestion,
            results: List<SearchResult>,
            responseInfo: ResponseInfo
        ) {
            Log.i("SearchApiExample", "Category search results: $results")
        }

        override fun onError(e: Exception) {
            Log.i("SearchApiExample", "Search error", e)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityRegisterBinding.inflate(layoutInflater)
        setContentView(binding?.root)

        searchEngine = MapboxSearchSdk.getSearchEngine()

        binding?.registerAddressId?.threshold = 3
        binding?.registerAddressId?.doOnTextChanged { text, _, _, _ ->
            searchRequestTask = searchEngine.search(
                text?.trim().toString(),
                SearchOptions(limit = 3),
                searchCallback
            )
        }

        binding?.registerCancelBtnId?.setOnClickListener { finish() }
        binding?.registerRegisterBtnId?.setOnClickListener { performAuth() }
        binding?.registerSelectProfileId?.setOnClickListener {
            isProfile = true
            selectImage()
        }
        binding?.btnSelectLegalId?.setOnClickListener {
            isProfile = false
            selectImage()
        }
        binding?.radioGroupId?.setOnCheckedChangeListener { _, checkedId ->
            when (checkedId) {
                R.id.rb_student_id -> userType = User.TYPE_STUDENT
                R.id.rb_volunteer_id -> userType = User.TYPE_VOLUNTEER
            }
        }

    }

    private fun performAuth() {
        val email = binding?.registerEmailId?.text.toString().trim()
        val password = binding?.registerPasswordId?.text.toString().trim()
        val firstName = binding?.registerFirstnameId?.text.toString().trim()
        val lastName = binding?.registerLastnameId?.text.toString().trim()
        val address = binding?.registerAddressId?.text.toString().trim()
        val phoneNumber = binding?.registerPhoneId?.text.toString().trim()

        if (!this::currentProfilePhotoUri.isInitialized) {
            "Please select a profile picture".showShortToast(this)
        }
        if (!this::currentLegalPhotoUri.isInitialized) {
            "Please select a picture of national id".showShortToast(this)
        }

        if (TextUtils.isEmpty(firstName)) {
            binding?.registerFirstnameId?.error = "FirstName is required"
            return
        }
        if (TextUtils.isEmpty(lastName)) {
            binding?.registerLastnameId?.error = "LastName is required"
            return
        }
        if (TextUtils.isEmpty(email)) {
            binding?.registerEmailId?.error = "Email is required"
            return
        }
        if (!email.matches(Patterns.EMAIL_ADDRESS.toRegex())) {
            binding?.registerEmailId?.error = "Email is badly formatted"
            return
        }
        if (TextUtils.isEmpty(password)) {
            binding?.registerPasswordId?.error = "Password is required"
            return
        }
        if (TextUtils.isEmpty(address)) {
            binding?.registerAddressId?.error = "Address is required"
            return
        }
        if (password.length < 8) {
            binding?.registerPasswordId?.error = "Password must be >= 8 characters"
            return
        }
        if (phoneNumber.length != 10) {
            binding?.registerPhoneId?.error = "Phone number must be equal to 10 digits"
            return
        }

        val dialog = ProgressDialog(this)
        dialog.setMessage("Registering your details, please wait...")
        dialog.setCanceledOnTouchOutside(false)
        dialog.show()

        val user = User(
            firstName = firstName,
            lastName = lastName,
            email = email,
            password = password,
            phoneNumber = phoneNumber,
            address = address,
            latitude = latitude,
            longitude = longitude,
            type = userType
        )

        val createUserTask = Firebase.auth.createUserWithEmailAndPassword(email, password)

        // register -> upload images -> get image urls -> push user details

        dialog.setMessage("Creating user account...")
        createUserTask.continueWithTask {
            if (it.isSuccessful) {
                dialog.setMessage("Uploading profile image...")
                Firebase.auth.currentUser?.uid?.let { userID ->
                    Firebase.storage
                        .reference
                        .child("images/${userID}/${user.firstName}-profile")
                        .putFile(
                            currentProfilePhotoUri,
                            storageMetadata { contentType = "image/jpeg" })
                }
            } else {
                "Error ! ${it.exception?.message}".showShortToast(this@RegisterActivity)
                null
            }
        }.continueWithTask {
            if (it.isSuccessful) {
                Firebase.auth.currentUser?.uid?.let { userID ->
                    Firebase.storage
                        .reference
                        .child("images/${userID}/${user.firstName}-profile")
                        .downloadUrl
                }
            } else {
                "Error ! ${it.exception?.message}".showShortToast(this@RegisterActivity)
                null
            }
        }.continueWithTask {
            if (it.isSuccessful) {
                user.profileImageLocation = it.result.toString()
                dialog.setMessage("Uploading ID documents...")
                Firebase.auth.currentUser?.uid?.let { userID ->
                    Firebase.storage
                        .reference
                        .child("images/${userID}/${user.firstName}-id")
                        .putFile(
                            currentLegalPhotoUri,
                            storageMetadata { contentType = "image/jpeg" })
                }
            } else {
                "Error ! ${it.exception?.message}".showShortToast(this@RegisterActivity)
                null
            }
        }.continueWithTask {
            if (it.isSuccessful) {
                Firebase.auth.currentUser?.uid?.let { userID ->
                    Firebase.storage
                        .reference
                        .child("images/${userID}/${user.firstName}-id")
                        .downloadUrl
                }
            } else {
                "Error ! ${it.exception?.message}".showShortToast(this@RegisterActivity)
                null
            }
        }.continueWithTask {
            if (it.isSuccessful) {
                user.idImageLocation = it.result.toString()
                dialog.setMessage("Uploading user details...")
                Firebase.auth.uid?.let { userID ->
                    user.id = userID
                    Firebase.firestore.collection("users")
                        .document(userID)
                        .set(user)
                }
            } else {
                "Error ! ${it.exception?.message}".showShortToast(this@RegisterActivity)
                null
            }
        }.addOnCompleteListener(this) { task ->
            if (task.isSuccessful) {
                "User Created".showShortToast(this@RegisterActivity)
                // sign out as creating account will sign in automatically
                Firebase.auth.signOut()
                finish()
            } else {
                if (task.exception?.message?.lowercase()?.contains("continuation") != true) {
                    "Error ! ${task.exception?.message}".showShortToast(this@RegisterActivity)
                }
            }
            dialog.dismiss()
        }
    }

    private fun selectImage() {
        val options = arrayOf("Take Photo", "Choose from Gallery", "Cancel")
        MaterialAlertDialogBuilder(this)
            .setTitle("Add Photo!")
            .setItems(options) { dialogInterface, i ->
                when {
                    options[i] == "Take Photo" -> {
                        dispatchTakePictureIntent()
                    }
                    options[i] == "Choose from Gallery" -> {
                        Intent(
                            Intent.ACTION_PICK,
                            MediaStore.Images.Media.EXTERNAL_CONTENT_URI
                        ).also {
                            startActivityForResult(it, IMAGE_PICK_CODE)
                        }
                    }
                    options[i] == "Cancel" -> {
                        dialogInterface.dismiss()
                    }
                }
            }.show()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == RESULT_OK) {
            if (isProfile == true) {
                if (requestCode == IMAGE_CAPTURE_CODE) {
                    binding?.registerimageSelectId?.load(currentProfilePhotoUri)
                } else if (requestCode == IMAGE_PICK_CODE) {
                    binding?.registerimageSelectId?.load(data?.data.also {
                        if (it != null) {
                            currentProfilePhotoUri = it
                        }
                    })
                }
            } else {
                if (requestCode == IMAGE_CAPTURE_CODE) {
                    binding?.ivLegalId?.load(currentLegalPhotoUri)
                } else if (requestCode == IMAGE_PICK_CODE) {
                    binding?.ivLegalId?.load(data?.data.also {
                        if (it != null) {
                            currentLegalPhotoUri = it
                        }
                    })
                }
            }
        }
    }

    private fun dispatchTakePictureIntent() {
        Intent(MediaStore.ACTION_IMAGE_CAPTURE).also { takePictureIntent ->
            takePictureIntent.resolveActivity(packageManager)?.also {
                val photoFile: File? = try {
                    createImageFile()
                } catch (ex: IOException) {
                    null
                }
                photoFile?.also {
                    val photoURI: Uri = FileProvider.getUriForFile(
                        this,
                        "com.example.rider.fileprovider",
                        it
                    )
                    takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI)
                    startActivityForResult(takePictureIntent, IMAGE_CAPTURE_CODE)
                }
            }
        }
    }

    private fun createImageFile(): File {
        val timeStamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault()).format(Date())
        val storageDir: File? = getExternalFilesDir(Environment.DIRECTORY_PICTURES)
        return File.createTempFile(
            "${if (isProfile == true) "PROFILE" else "ID"}_JPEG_${timeStamp}_",
            ".jpg",
            storageDir
        ).apply {
            if (isProfile == true) {
                currentProfilePhotoUri = Uri.fromFile(File(absolutePath))
            } else {
                currentLegalPhotoUri = Uri.fromFile(File(absolutePath))
            }
        }
    }

    override fun onDestroy() {
        searchRequestTask.cancel()
        super.onDestroy()
    }

    companion object {
        const val TAG = "RegisterActivity"
        private const val IMAGE_PICK_CODE = 1000
        private const val IMAGE_CAPTURE_CODE = 1001
    }
}

